File: skill-test/projects/sample-cluster-java/src/main/java/cluster/java/cafe24/Cafe24MetricsSingletonActor.java
package cluster.java.cafe24;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

public class Cafe24MetricsSingletonActor extends AbstractActor {

    public record RecordCall(String mallId, int statusCode, long queueDelayMs) implements Serializable {
    }

    public record GetMallMetrics(String mallId) implements Serializable {
    }

    public record MallMetrics(String mallId, long totalCalls, long throttled429, double avgQueueDelayMs)
        implements Serializable {
    }

    public enum Stop implements Serializable {
        INSTANCE
    }

    private static final class MetricState {
        long totalCalls;
        long throttled429;
        long sumQueueDelayMs;
    }

    private final Map<String, MetricState> stateByMall = new HashMap<>();

    public static Props props() {
        return Props.create(Cafe24MetricsSingletonActor.class, Cafe24MetricsSingletonActor::new);
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
            .match(RecordCall.class, this::onRecordCall)
            .match(GetMallMetrics.class, this::onGetMallMetrics)
            .match(Stop.class, stop -> getContext().stop(getSelf()))
            .build();
    }

    private void onRecordCall(RecordCall msg) {
        MetricState state = stateByMall.computeIfAbsent(msg.mallId(), key -> new MetricState());
        state.totalCalls += 1;
        if (msg.statusCode() == 429) {
            state.throttled429 += 1;
        }
        state.sumQueueDelayMs += msg.queueDelayMs();
    }

    private void onGetMallMetrics(GetMallMetrics msg) {
        MetricState state = stateByMall.getOrDefault(msg.mallId(), new MetricState());
        double avgDelay = state.totalCalls == 0 ? 0.0 : (double) state.sumQueueDelayMs / state.totalCalls;
        getSender().tell(
            new MallMetrics(msg.mallId(), state.totalCalls, state.throttled429, avgDelay),
            getSelf());
    }
}

File: skill-test/projects/sample-cluster-dotnet/ClusterActors/Cafe24MetricsSingletonActor.cs
using Akka.Actor;

namespace ClusterActors;

public sealed class Cafe24MetricsSingletonActor : ReceiveActor
{
    public sealed record RecordCall(string MallId, int StatusCode, long QueueDelayMs);
    public sealed record GetMallMetrics(string MallId);
    public sealed record MallMetrics(string MallId, long TotalCalls, long Throttled429, double AvgQueueDelayMs);
    public sealed record Stop;

    private sealed class MetricState
    {
        public long TotalCalls { get; set; }
        public long Throttled429 { get; set; }
        public long SumQueueDelayMs { get; set; }
    }

    private readonly Dictionary<string, MetricState> _stateByMall = new();

    public Cafe24MetricsSingletonActor()
    {
        Receive<RecordCall>(msg =>
        {
            if (!_stateByMall.TryGetValue(msg.MallId, out var state))
            {
                state = new MetricState();
                _stateByMall[msg.MallId] = state;
            }

            state.TotalCalls += 1;
            if (msg.StatusCode == 429) state.Throttled429 += 1;
            state.SumQueueDelayMs += msg.QueueDelayMs;
        });

        Receive<GetMallMetrics>(msg =>
        {
            if (!_stateByMall.TryGetValue(msg.MallId, out var state))
            {
                Sender.Tell(new MallMetrics(msg.MallId, 0, 0, 0));
                return;
            }

            var avg = state.TotalCalls == 0 ? 0 : (double)state.SumQueueDelayMs / state.TotalCalls;
            Sender.Tell(new MallMetrics(msg.MallId, state.TotalCalls, state.Throttled429, avg));
        });

        Receive<Stop>(_ => Context.Stop(Self));
    }
}
